src/script_interface/packed_variant.hpp

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (C) 2020 The ESPResSo project
+ *
+ * This file is part of ESPResSo.
+ *
+ * ESPResSo is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ESPResSo is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef SCRIPT_INTERFACE_PACKED_VARIANT_HPP
+#define SCRIPT_INTERFACE_PACKED_VARIANT_HPP
+
+#include "Variant.hpp"
+
+#include <functional>
+#include <unordered_map>
+#include <utility>
+
+namespace ScriptInterface {
+using ObjectId = std::size_t;
+
+/**
+ * @brief Id for object.
+ *
+ * This assigns every ObjectHandle a unique id.
+ */
+inline ObjectId object_id(const ObjectHandle *p) {
+  static_assert(sizeof(const ObjectHandle *) <= sizeof(ObjectId), "");
+  return reinterpret_cast<ObjectId>(p);
+}
+
+/**
+ * @brief Variant value with references replaced by ids.
+ */
+using PackedVariant = boost::make_recursive_variant<
+    None, bool, int, double, std::string, std::vector<int>, std::vector<double>,
+    ObjectId, std::vector<boost::recursive_variant_>, Utils::Vector2d,
+    Utils::Vector3d, Utils::Vector4d>::type;
```

Comment:
Currently, `ScriptInterface::Variant` can only hold object ids, with this, it can hold shared pointers to objects. This is more covenient, because the `Variant` keeps the pointee alive and no special treatment is needed for them. The packed version is needed because for sending over mpi and for serialization we can't store pointers. In this case the pointers are replaced by ids and the client has to somehow deal with the shared ownership, and provide pointers again when converting back.

Reply:
ok I see now what you meant (`PackedVariant` contains an `ObjectId` while `Variant` contais an `ObjectRef`), but it's not intuitive and very easy to overlook that difference. We need to write this down in some stand-alone doxygen file. For example, If we want to add support for a new type like `Vector3i` we need to touch both `PackedVariant` and `Variant`. Since `ObjectId` is a typedef to `size_t`, does this prevent us from adding `size_t` in the variants due to type collision?

Reply:
What I really would like to have is some notion of native types (as opposed to object handles), which are the same in the packed and unpacked version. But because all the variant stuff is used all over the client code at the moment it is too cumbersome to change that. I also don't think that new types should be added to the list of native types, that is not
a scalable solution. I'm not a fan of technical writing in some doxygen  file, that will just get out of sync with the code,
at the end it can't be avoided to read the code anyway to understand how the library code works.

Reply:
Also the `PackedVariant` is some internal technical detail, even if clients roll their own `ObjectHandle` implementation (which they shouldn't), they don't have to know about this.

Reply:
> I'm not a fan of technical writing in some doxygen file, that will just get out of sync with the code

Doxygen is designed to detect when such technical files get out of sync with the code. You can even move the docstrings out of the code and into stand-alone `.dox` files if you want to, as long as you provide the function/class/variable name, it will generate valid documentation. This concern you have is only the result of contributors writing stand-alone doxygen files *without* adding `@ref` in front of every function name, which means there is no way for them to detect desynchronized docs by reading the doxygen warnings and no way for CI to complain about it. I actually had to create a dictionary of function names for my espresso spellchecker because the parser rarely finds a `@ref` in front of non-english words.

> at the end it can't be avoided to read the code anyway to understand how the library code works

We don't want to be in a situation where we have to read the library code to understand how it works. It took me months to create a RDF observable because I had to understand the entire script interface. This was a tremendous waste of my time, because everything I learned will become useless once we merge this PR.

Reply:
> It took me months to create a RDF observable because I had to understand the entire script interface. 

Frankly I don't think this is true, others managed to add classes without reading all the code. I agree that the client side should be simplified thou, but I don't this will happen.